# Profitable App Profiles for the App Store and Google Play Markets
This project highlights how to perform data analysis for a company that builds Android and iOS mobile apps. The apps are available on Google Play and in the App Store.

The company only builds apps that free to download and install, and their main source of revenue consists of in-app ads. The main goal for this project is to analyze data to help our developers find mobile app profiles that are profitable for the App Store and Google Play markets.

One thing that is notable for this company is that the revenue for any given app is mostly influenced by the number of users that use that app. Therefore in this project, the main goal is to analyze data to help our developers understand what kinds of apps are likely to attract more users.

Below we open the two data sets and then continue with exploring the data.

In [29]:
from csv import reader

### The Google Play data set ###
opened_file = open('googleplaystore.csv', encoding='utf8')
read_file = reader(opened_file)
google = list(read_file)
google_header = google[0]
google_list = google[1:]

### The App Store data set ###
opened_file = open('AppleStore.csv', encoding='utf8')
read_file = reader(opened_file)
apple = list(read_file)
apple_header = apple[0]
apple_list = apple[1:]

To make it easier to explore the two data sets, we use the function below named `explore_data` that is used repeatedly to explore rows in a more readable way. The function also has an option to show the number of rows and columns for any data set.

In [34]:
# The function that displays the contents of a given dataset with option to specify start and end rows and number of rows and columns
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        #print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

## Exploring the IOS App Store Dataset
This is the [link](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) detailing the description of the column names for the IOS App Store dataset.

The IOS App store contains 7197 apps and 16 columns. Some important column names may be 'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', and 'prime_genre'.

In [36]:
# Exploring the IOS App Store dataset
print(apple_header)
print()
explore_data(apple_list, 0, 2, rows_and_columns=True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']
['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']
Number of rows: 7197
Number of columns: 16


## Exploring the Google Playstore Dataset
This is the [link](https://www.kaggle.com/lava18/google-play-store-apps) detailing the description for the column names for the Google Play Store dataset.

The Google Play data set has 10841 apps and 13 columns. Some important columns that might be useful for the purpose of our analysis are 'App', 'Category', 'Reviews', 'Installs', 'Type', 'Price', and 'Genres'.

In [37]:
# Printing the header row for the Google Play Store dataset
print(google_header)
print()
explore_data(google_list, 0, 2, rows_and_columns=True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']
['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']
Number of rows: 10841
Number of columns: 13


# Deleting Wrong Data
The function `row_check` below checks for missing columns by checking the length of each row in a dataset. It takes in a `data_set` list and `store` name whether `google` or `apple`. The function performs the following:
- checks to see if the store is either google or apple
- for each row in dataset checks its length
- comapares each row length to known length, 13 for google and 16 for apple
- for each row, if row length not the same as given number, the function prints out that row and its index


In [38]:
# Function to check for missing columns in dataset
def row_check(data_set_list, store):

    selected_store = ''
    length = 0

    if store == 'google':
        selected_store, length = 'google', 13
    elif store == 'apple':
        selected_store, length = 'apple', 16
    else:
        return print('Wrong App Store Name')

    for row in data_set_list:
        if len(row) != length:
            print(row)
            print(data_set_list.index(row))

**Checking the rows with missing columns**<br>
To make it easier to explore the two data sets, we use the function below named `explore_data` that is used repeatedly to explore rows in a more readable way. The function also has an option to show the number of rows and columns for any data set.

Looking at the the row that has been extracted, it shows the number 19 for rating value (column 3) which is not correct as the rating values can only be upto 5. This shows that this row is ceratinly missing something and will hence be removed.

In [39]:
# Checking missing columns in google_list and apple_list
row_check(google_list, 'google')
row_check(apple_list, 'apple')

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10472


**Deleting the row with missing column in the google_list**

In [40]:
print(len(google_list))
del google_list[10472]  # don't run this more than once
print(len(google_list))

10841
10840


# Removing Duplicate Entries
## Part One

Upon exploring the Google Play data set, its evident that there are some some apps that have more than one entry. Like for example checking for the slack app below shows the Slack app appears three times in the dataset.

In [41]:
for app in google_list:
    name = app[0]
    if name == 'Slack':
        print(app)

['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


The code below checks whether there are duplicates in the `google_list`<br>
The functions performs the following:
- for each app in the google list, creates two empty lists named `store_unique_apps` to store uniques apps and `store_duplicate_apps` to store duplicate apps
- the code loops through the `google_list` checking each app
- if an app name is not in the `store_unique_apps` list, its included in `store_unique_apps` list otherwise its included in the `store_duplicate_apps`
- finally the code prints out the number of unique and duplicate apps and some expmples of the duplicate apps

In [45]:
# The code below checks whether there are duplicates in the google_list
google_unique_apps = [] 
google_duplicate_apps = [] 

for app in google_list:
        app_name = app[0]
        
        if app_name not in google_unique_apps:
           google_unique_apps.append(app_name)
        else:
            google_duplicate_apps.append(app_name)

print('Number of unique apps:', len(google_unique_apps))
print('Number of duplicate apps:', len(google_duplicate_apps))
print('Examples of duplicate apps:', google_duplicate_apps[:15]) # Only show 15 example apps

Number of unique apps: 9659
Number of duplicate apps: 1181
Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


We see that there are 1181 duplicate apps in this `google_list` dataset

# Part Two
Examining the `google_list`, specifically for `slack` app its evident that there are duplicates in the google_list for this app. The only difference is in the fourth position of each row, which indicates the number of reviews. The different numbers show the data was collected at different times.

This information is used to build a criterion for removing the duplicates whereby the higher the number of reviews, the more recent the data should be. Only the row with the highest number of reviews is to be kept and the other entries for any given app are removed as such.

In [16]:
# This code checks each app in the google_list dataset and only keeps the apps with the highest 
# number of reviews where there are duplicate apps appearing in the dataset
reviews_max = {}
for app in google_list:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

Because there are 1181 cases of duplicate app occurences, the length of unique dictionary should be equal to the difference between the length of the data set and 1181.

In [17]:
print(len(google_list)-len(google_duplicate_apps))
print(len(reviews_max))

9659
9659


This section makes use of the reviews_max dictionary to remove the duplicates. For each of the duplicates, only the entry with the highest number of reviews are to be kept. The following steps are perfomed to achieve this:

- First create two empty lists, google_clean and already_added.
- Loop through google_list data set, and for every iteration:
   - Extract each app name and number of reviews.
   - Add current row (app) to android_clean list, and app name (name) to already_added list if:
     - The number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary; and
     - The name of the app is not already in the already_added list.
- This supplementary condition is added to ensure that those cases where the highest number of reviews of a duplicate app is the same for more than one entry. Only checking for reviews_max(name) == n_reviews, will end up with duplicate entries for some apps.



In [18]:
# This code cleans the google_list dataset by removing the duplicate apps with less reviews
google_clean = []
already_added = []
for app in google_list:
    name = app[0]
    n_reviews = float(app[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        google_clean.append(app)
        already_added.append(name)

To confirm the `google_list` dataset has been cleaned, the `google_clean` dataset is explored using the `explore_data` function. The dataset has 9659 rows as expected.

In [46]:
explore_data(google_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']
['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']
['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']
Number of rows: 9659
Number of columns: 13


## Confirming no duplicates in the apple_list

In [44]:
# This code check for duplicate apps in the apple_list dataset
apple_unique_apps = [] 
apple_duplicate_apps = [] 

for app in apple_list:
        app_name = app[0]
        
        if app_name not in apple_unique_apps:
            apple_unique_apps.append(app_name)
        else:
            apple_duplicate_apps.append(app_name)

print('Number of unique apps:', len(apple_unique_apps))
print('Number of duplicate apps:', len(apple_duplicate_apps))
print('Examples of duplicate apps:', apple_duplicate_apps[:15]) # Only show 15 example apps

Number of unique apps: 7197
Number of duplicate apps: 0
Examples of duplicate apps: []


# Removing Non-English Apps
## Part One

The company that we are doing this analysis for is only interested in apps that are designed for an English-speaking audience. Therefore, this part seeks to explore and remove apps in the dataset that are designed for non-English speaking people.

In [17]:
# Here we explore some examples of apps with non-English characters
print(apple_list[813][1])
print(apple_list[6731][1])
print()
print(google_clean[4412][0])
print(google_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜

中国語 AQリスニング
لعبة تقدر تربح DZ


The numbers corresponding to the characters commonly used in English text are in the range of 0 to 127 according to the ASCII (American Standard Code for Information Interchange) system. A criteria based on this number range is used to build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.

If an app name contains a character that is greater than 127, then it be that that the app has a non-English name. The criteria of string index and iterating over each string index to detect non-English apps is used for this part.

In the following section, we develop the above described function. 

In [18]:
# Function is_english checks for apps with charatcters in the app name that are not English characters
# The functions returns False for apps that have characters in their name that are not English characters
# Otherwise it returns True
def is_english(string):
    for character in string:
        if ord(character) > 127:
            return False
        return True

In code cell below, we check some app names if they are English apps or not.

In [19]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
True
True


Even if the function correctly identifies the apps above, we still need to ensure that we don't mis-classify apps, so we need to add something to our function.

To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This means all English apps with up to three emoji or other special characters will still be labeled as English.

In [22]:
def is_english_new(string):
    non_english = 0
    
    for character in string:
        if ord(character) > 127:
            non_english += 1
    
    if non_english > 3:
        return False
    else:
        return True

In [23]:
print(is_english_new('Docs To Go™ Free Office Suite'))
print(is_english_new('Instachat 😜'))
print(is_english_new('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Now we use this function to filter out all non-English apps from both datatsets.

In [48]:
# Checking the number of Google Playstore English apps remaining after 
# filtering using the is_english_new function
google_english = []
for app in google_clean:
    if is_english_new(app[0]) == True:
        google_english.append(app)

explore_data(google_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']
['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']
['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']
Number of rows: 9614
Number of columns: 13


In the Google Play dataset, we remain with 9614 Engish apps out of a total of 9659 apps after filtering out the non-English apps.

In [49]:
# Checking the number of IOS Store English apps remaining after 
# filtering
apple_english = []
for app in apple_list:
    if is_english_new(app[1]) == True:
        apple_english.append(app)

explore_data(apple_english, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']
['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']
['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']
Number of rows: 6183
Number of columns: 16


In the IOS dataset, we remain with 6183 Engish apps out of a total of 7197 apps after filtering out the non-English apps.


# Isolating the Free Apps

In this analyisis, we are only interested in apps that are free to download and install and the main source of revenue consists of in-app ads. The two datasets both contain free and non-free apps and hence we need to isolate only the free apps for our analysis.

This step of isolating the free apps is the final step in the data cleaning process before we start on the data analysis.

In [50]:
# Isolating the free Englsih Google Playstore apps
google_free = []
for app in google_english:
    if app[7] == '0': # Column 7 stores the price information in the Google dataset
        google_free.append(app) # We populate the google_free dataset with free apps only

explore_data(google_free, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']
['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']
['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']
Number of rows: 8864
Number of columns: 13


In [51]:
# Isolating the free Englsih IOS Store apps
apple_free = []
for app in apple_english:
    if app[4] == '0.0':
        apple_free.append(app)

explore_data(apple_free, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']
['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']
['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']
Number of rows: 3222
Number of columns: 16


At this point, we have performed the following data cleaning steps:

- Removing inaccurate data
- Removing duplicate app entries
- Removing non-English apps
- Isolating the free apps

After all above steps, we are left with 8864 Android apps and 3222 iOS apps for our analysis.

# Most Common Apps by 
## Part One

Because the number of people using our apps affect our revenue, the main aim of this analysis is to determine the apps that are likely to attract more users.

To minimize risks and overhead, the validation strategy we will use for an app idea has three steps:

- Build a minimal Android version of the app and add it to Google Play.
- If the app has a good response from users, we develop it further.
- If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful in both markets.

This part of analysis begins with determining the most common genres for each market. This is achieved by building a frequency table for the prime_genre column of the App Store data set and the Genres and Category columns of the Google Play data set.

## Part Two

Below we develop the two functions that will be used to analyze the frequency tables:

- One function to generate frequency tables that show percentages
- Another function to display the percentages in a descending order

In [54]:
# This function generates frequency tables that show percentages
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages # Frequency table with percentages

In [55]:
# This function displays the percentages in a descending order
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Displaying the frequency table for the `prime_genre` column of the IOS Store dataset

In [28]:
display_table(apple_free, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


It is evident that among English free iOS apps, 58.2% are games. Entertainment apps are 7.9%, followed by Photo and Video apps (5.0%), then Education apps (3.7%). Social Networking apps account for 3.3% of the apps in our data set.

Overall, for free English apps, App Store is dominated by apps that are designed for fun (games, entertainment, social networking, Photo and Video, etc.). While apps with practical purposes (shopping, utilities, sports, productivity, health & fitness, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

Displaying the frequency table for the `Genres` column of the Google Playstore dataset.

In [29]:
display_table(google_free, 9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

Displaying the frequency table for the `Category` column of the Google Playstore dataset.

In [30]:
display_table(google_free, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

For the Google Playstore dataset, the difference between the Genres and the Category columns is not so clear. Although we can notice that the Genres column has more categories. We will only work with the Category column moving forward since we are only interested in the big picture.

Our conclusion from analysis so far is that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

# Most Popular Apps by Genre on the App Store

We want to find out the genres that are more popular or have the most users and we do this by calculating the average number of installs for each app genre. The Google Play data set shows this information in the olumn named Installs but the IOS App Store dataset is missing this information. Instead we can use the total number of user ratings as a substitue which is found in the rating_count_tot column.

We calculate the average number of user ratings per app genre on the App Store. We follow the following steps:

- Isolate the apps of each genre
- Add up the user ratings for the apps of that genre
- Divide the sum by the number of apps belonging to that genre (not by the total number of apps)

We calculate the average number of user ratings for each genre by:
- Using a for loop to iterate inside another for loop.

In [56]:
# Code used to calculate average number of user ratings per app genre on the App Store
genres_ios = freq_table(apple_free, -5)

for genre in genres_ios:
    total = 0
    genre_len = 0
    for app in apple_free:
        app_genre = app[-5]
        if app_genre == genre:
            n_ratings = float(app[5])
            total += n_ratings
            genre_len += 1
    avg_ratings = total / genre_len
    print(genre, ':', avg_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


From the analyis of the average ratings for the IOS App Store dataset, its is clear that profile of Navigation apps, with an average rating of about 86,090.33, are the most popular apps going by the user ratings. This is motly dominated by two main contenders, Waze and Google Maps which account for about half a million total number of ratings.

In [57]:
for app in apple_free:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


In [58]:
for app in apple_free:
    if app[-5] == 'Social Networking':
        print(app[1], ':', app[5]) # print name and number of ratings

Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778
MeetMe - Chat and Meet New People : 97072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90414
InsTrack for Instagram - Analytics Plus More : 85535
Tango - Free Video Call, Voice and Chat : 75412
LinkedIn : 71856
Match™ - #1 Dating App. : 60659
Skype for iPad : 60163
POF - Best Dating App for Conversations : 52642
Timehop : 49510
Find My Family, Friends & iPhone - Life360 Locator : 43877
Whisper - Share, Express, Meet : 39819
Hangouts : 36404
LINE PLAY - Your Avatar World : 34677
WeChat : 34584
Badoo - Meet New People, Chat, Socialize. : 34428
Followers + for Instagram - Follower Analytics : 28633
GroupMe : 28260
Marco Polo Video Walkie Talkie : 27662
Miitomo : 2

Similary for Social Networking apps, the majority of ratings are those for Facebook (2,974,676), Pinterest (1,061,624) and Skype for iPhone (373,519)

In [34]:
for app in apple_free:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5]) # print name and number of ratings

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


For the apps with genre Reference (Average of 74,942), the most popular apps are those for Bible and Dictionaries. which actually skews up the average rating.

The Bible genre offers some promise. An option could be to take another popular book and turn it into an app where we could add different features besides the raw version of the book such as daily quotes from the book, an audio version of the book, quizzes about the book, inside dictionary, etc.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Here we explore the number of installs for apps in the Google Playstore. The information, howver, does not seem to be precise enough. It only gives an indication of the number of installs and is not the actual number of installs. But we can still use this information to estimate the number of installs wjich will be siffcient for our case. Most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [35]:
display_table(google_free, 5) # the Installs columns

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


For our analysis, we need to convert each install number to float and remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [36]:
# Here we check the Google Play Store for apps with high numbers of installs
category_android = freq_table(google_free, 1)

for category in category_android:
    total = 0
    category_len = 0
    for app in google_free:
        app_category = app[1]
        if app_category == category:
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            category_len += 1
    avg_installs = total / category_len
    print(category, ':', avg_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, Communication apps have the most installs with 38,456,119 average installs. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [37]:
for app in google_free:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we removed all the communication apps that have over 100 million installs, the average is reduced roughly ten times:

In [38]:
under_100_m = []

for app in google_free:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

3603485.3884615386

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:

In [39]:
for app in google_free:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [40]:
for app in google_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [41]:
for app in google_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

The app profile falling in the software for processing and reading ebooks category including collections of libraries and dictionaries may not be idea and hence would not be a good idea to build similar apps since there'll be some significant competition.

Another thing to note is that because they are quite a few apps built around the book Quran, this may sugest that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

To make this app more appealing, we may have to add special features besides the raw version of the book, such as daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

# Conclusion

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.